### ISSUE TITLE
When using `cy.json`, sometimes some nodes are not being deleted

### ISSUE THREAD

**Issue by josejulio** [cytoscape/cytoscape.js#2231]
**Issue type**

Bug report

**Environment info**

- Cytoscape.js version 3.2.22 (i suppose it might happen on 3.3.0 too):
- Browser/Node.js & version : Chrome Version 70.0.3538.67 (Official Build) (64-bit)

**Current (buggy) behaviour**

When using `cy.json`, sometimes some nodes are not being deleted

**Desired behaviour**

When using `cy.json`, all nodes not in the provided json should be deleted.

**Comment by josejulio** [comment]
I think the bug was introduced here:

https://github.com/cytoscape/cytoscape.js/commit/47a354eba7d5143ceb3778446921bdd3ae3924d1#diff-4cbda7cf5642b53f0b4ada2eb9e65229R339


The for goes from 0 to eles.length, but the content of the loop is removing elements, thus updating eles.length.

I'm working on the bug, but we should either go from eles.length to 0, or filter all the nodes that are going to be deleted and iterate over that array.

**Comment by maxkfranz** [comment]
Removing from the graph doesn't affect the contents of a collection.  Do you have a reproducible test case?

**Comment by josejulio** [comment]
Is reproducible in the context of the application, but couldn't reproduce in a jsbin.

Let me try again.

**Comment by josejulio** [comment]
Also, not sure if this helps, but the collection used is the `mutableElements` defined in `core/search.js` as:
```js
mutableElements: function(){
    return this._private.elements;
  }
```

**Comment by josejulio** [comment]
got this reproducer: https://jsbin.com/qulepikabe

Click the cat and it will remove about the half, when it should remove all but one.

**Comment by josejulio** [comment]
Smaller one, which might be better for unit testing. I'm going to add to my PR:
https://jsbin.com/muyebuhana

**Comment by josejulio** [comment]
Is `3.3.1` scheduled for any day of this week? I'm wondering to know how to deal with an issue, thank you!

**Comment by maxkfranz** [comment]
I've been away and unable to release.  I'll publish tomorrow morning.

**Comment by josejulio** [comment]
Thank you
